### DISCUSSION TITLE
FhirValue - Dosage rendering ?

### DISCUSSION THREAD

**Discussion by jy95** [bonfhir/bonfhir#284] [Q&A]
Hello,

First thing first, thanks for your project : it is quite clean and powerful.
I read your [documentation](https://bonfhir.dev/packages/react/components/fhir-value) / [Storybook](https://bonfhir.dev/storybook/?path=/docs/bonfhir-data-display-fhirvalue--docs) and I did noticed that you don't have a Storybook representing the [Dosage](https://build.fhir.org/dosage.html) datatype. Is it planned at some point in the future ?

I have my own library atm for that : [fhir-dosage-utils](https://github.com/jy95/fhir-dosage-utils) . I would be curious to see what would be the approach of BonFHIR

Thanks in advance

**Comment by opiation** [comment]
Hey there! Thanks for reaching out and sharing! We don't have a published roadmap as of yet but we're working on it. Nonetheless, this seems like an entirely reasonable request.

To support this, I think we'd need to add one or more value formatters for the `Dosage` type in `@bonfhir/core`, test support in the `@bonfhir/react` `<FhirValue>` and add an example in the Storybook. @julienblin, @lp, does this sound sensible? If so, I'll create an issue for it.

I can't promise any kind of time frame but I will take a look at your library for string representations of dosages that would be worthwhile supporting with `<FhirValue type="Dosage">`.

**Reply by opiation** [reply]
For some additional context, most presentation of FHIR data types and primitives is handled by [ValueFormatter](https://github.com/bonfhir/bonfhir/blob/7fb3b15ebf3b9485dae7b9f167441179446054ce/packages/core/src/r5/formatters.ts#L49)s labeled with a `type` string name. This allows `<FhirValue ... />` and other utilities to select appropriate formatters. Thus, to support `Dosage`, we'd add one more formatters like so:

```ts
type DosageFormatOptions = {
  // Options used for the `format` function to alter how/what dosage information gets displayed
};

const dosageFormatter: ValueFormatter<"Dosage", Dosage | null | undefined, DosageFormatOptions> = {
  type: "Dosage",
  format: (dosageValue, _, opts) => {
    // ...here be core formatting logic dragons
  }
}
```

Here's the [`Reference` formatter](https://github.com/bonfhir/bonfhir/blob/main/packages/core/src/r5/value-formatters/reference.ts) as an example.

I've not yet looked at `fhir-dosage-utils` thoroughly but at a quick glance, I think the immediate features that seem most easily portable to `@bonfhir/core` would be the helpers under `src/utils` and the methods of `FhirDosageUtils`.
